File: examplesForThesis-R/ch4-example-gp-acqfunc.R
########################################################################
#
# Code to reproduce examples from the Licentiate's thesis:
# Sequential Monte Carlo for inference in nonlinear state space models
#
# written by
# Johan Dahlin ( johan.dahlin (at) liu.se
#
# Available from:
# http://users.isy.liu.se/en/rt/johda87/
#
# Copyright (c) 2014 Johan Dahlin
# 
# Run this code together with the corresponding R-file to reproduce:
#
# GPO using different acquisition rules
# Example 4.7 in Section 4.4.2 
#
########################################################################

# Helper function to normalise the AQs
normalise <- function(x){
  return( x/max(x) - min(x)/max(x) );
}


dat   = read.table("ch4-example-gp-acqfunc-data.csv",sep=",",header=T)[,-1]
datr  = read.table("ch4-example-gp-acqfunc-datared.csv",sep=",",header=T)[,-1]
gp3   = read.table("ch4-example-gp-acqfunc-3d.csv",sep=",",header=T)[,-1]
gp5   = read.table("ch4-example-gp-acqfunc-5d.csv",sep=",",header=T)[,-1]
gp10  = read.table("ch4-example-gp-acqfunc-10d.csv",sep=",",header=T)[,-1]

layout(matrix(1:6, 3, 2, byrow = FALSE)); par(mar=c(4,5,1,1)) 

plot(dat[,1],gp3[,1],type="l",lwd=3,ylim=c(-8,30),xlim=c(0.30,9.65),xlab="x",ylab="f(x)"); 
polygon( c(dat[,1],rev(dat[,1])), c(gp3[,3], rev(gp3[,2])), col="lightgrey", border=NA)
lines(dat[,1],gp3[,1],lwd=3); lines(dat[,1],dat[,2],lwd=3,lty="dashed"); points(datr[1:3,1],datr[1:3,2],pch=19,cex=1.5)
text(10,30,"Matérn 5/2, N=3",pos=2);

plot(dat[,1],gp5[,1],type="l",lwd=3,ylim=c(-8,30),xlim=c(0.30,9.65),col="black",xlab="x",ylab="f(x)");
polygon( c(dat[,1],rev(dat[,1])), c(gp5[,3], rev(gp5[,2])), col="lightgrey", border=NA)
lines(dat[,1],gp5[,1],lwd=3,col="black"); lines(dat[,1],dat[,2],lwd=3,lty="dashed"); points(datr[1:5,1],datr[1:5,2],pch=19,cex=1.5)
text(10,30,"Matérn 5/2, N=5",pos=2);

plot(dat[,1],gp10[,1],type="l",lwd=3,ylim=c(-8,30),xlim=c(0.30,9.65),col="black",xlab="x",ylab="f(x)");
polygon( c(dat[,1],rev(dat[,1])), c(gp10[,3], rev(gp10[,2])), col="lightgrey", border=NA)
lines(dat[,1],gp10[,1],lwd=3,col="black"); lines(dat[,1],dat[,2],lwd=3,lty="dashed"); points(datr[1:10,1],datr[1:10,2],pch=19,cex=1.5)
text(10,30,"Matérn 5/2, N=10",pos=2);

plot(dat[,1],normalise(gp3[,3]),type="l",lwd=3,ylim=c(0,1.2),xlim=c(0.30,9.65),xlab="x",ylab="AQ(x)",col="darkgreen"); 
lines(dat[,1],normalise(gp3[,4]),lwd=3,col="darkred")
lines(dat[,1],normalise(gp3[,5]),lwd=3,col="steelblue")
text(10,0.85,"PI",pos=2,font=2,cex=1.2,col="darkgreen");
text(10,1.00,"UCB",pos=2,col="steelblue",font=2,cex=1.2);
text(10,0.35,"EI",pos=2,col="darkred",font=2,cex=1.2);

plot(dat[,1],normalise(gp5[,3]),type="l",lwd=3,ylim=c(0,1.2),xlim=c(0.30,9.65),xlab="x",ylab="AQ(x)",col="darkgreen");
lines(dat[,1],normalise(gp5[,4])-0.02,lwd=3,col="darkred")
lines(dat[,1],normalise(gp5[,5]),lwd=3,col="steelblue")

plot(dat[,1],normalise(gp10[,3]),type="l",lwd=3,ylim=c(0,1.2),xlim=c(0.30,9.65),xlab="x",ylab="AQ(x)",col="darkgreen");
lines(dat[,1],normalise(gp10[,4])-0.03,lwd=3,col="darkred")
lines(dat[,1],normalise(gp10[,5]),lwd=3,col="steelblue")

########################################################################
# End of file
########################################################################

File: examplesForThesis-python/ch4-example-gp-acqfunc.py
########################################################################
#
# Code to reproduce examples from the Licentiate's thesis:
# Sequential Monte Carlo for inference in nonlinear state space models
#
# written by
# Johan Dahlin ( johan.dahlin (at) liu.se
#
# Available from:
# http://users.isy.liu.se/en/rt/johda87/
#
# Copyright (c) 2014 Johan Dahlin
# 
# Run this code together with the corresponding R-file to reproduce:
#
# GPO using different acqusition rules
# Example 4.7 in Section 4.4.2 
#
########################################################################

import numpy as np
import scipy as sp
import pylab as pb
import GPy
import pandas

########################################################################
# Create the underlying function and evaluate it on a grid
########################################################################
x = np.linspace(0.,10.,500); x = x[:,None];
y = (3*np.cos(x/2) + np.sin(2+x/4) )**2;

# Select some random samples of the function and add some noise to the observations
xred = x[[ 20, 197, 317, 149,  89, 443,  53, 353, 395, 186, 292, 359, 326, 369,  99]]
yred = y[[ 20, 197, 317, 149,  89, 443,  53, 353, 395, 186, 292, 359, 326, 369,  99]] + np.sqrt(2) * np.random.normal(size=(15,1))

# Calculate the GPO model using all the data
k3 = GPy.kern.bias(1) + GPy.kern.Matern52(1, lengthscale=1);     
m3 = GPy.models.GPRegression(xred,yred,k3); m3.optimize();

########################################################################
# Fit the GP to the first 3 observations and predict the values of the grid
########################################################################

m3 = GPy.models.GPRegression(xred[0:3],yred[0:3],k3);
Mup, ys2, up95, lo95 = m3.predict( np.array(x) ); 

# Compute the values of the different AQs over the grid
mumax = max(Mup);
Z     = ( Mup - mumax - 0.01 ) / np.sqrt( ys2 )
PI    = sp.stats.norm.cdf( Z , 0, 1)
EI    = Z * sp.stats.norm.cdf( Z, 0, 1) + np.sqrt( ys2 ) * sp.stats.norm.pdf( Z, 0, 1)
UCB   = up95;

# Export data to R for plotting
out = np.hstack((Mup,up95,lo95,PI,EI,UCB));
pandas.DataFrame(out).to_csv("../figures/ch4-example-gp-acqfunc/ch4-example-gp-acqfunc-3d.csv");

########################################################################
# Fit the GP to the first 5 observations and predict the values of the grid
########################################################################

m3 = GPy.models.GPRegression(xred[0:5],yred[0:5],k3);
Mup, ys2, up95, lo95 = m3.predict( np.array(x) ); 

# Compute the values of the different AQs over the grid
mumax = max(Mup);
Z     = ( Mup - mumax - 0.01 ) / np.sqrt( ys2 )
PI    = sp.stats.norm.cdf( Z , 0, 1)
EI    = Z * sp.stats.norm.cdf( Z, 0, 1) + np.sqrt( ys2 ) * sp.stats.norm.pdf( Z, 0, 1)
UCB   = up95;

# Export data to R for plotting
out = np.hstack((Mup,up95,lo95,PI,EI,UCB));
pandas.DataFrame(out).to_csv("../figures/ch4-example-gp-acqfunc/ch4-example-gp-acqfunc-5d.csv");

########################################################################
# Fit the GP to the first 10 observations and predict the values of the grid
########################################################################

m3 = GPy.models.GPRegression(xred[0:10],yred[0:10],k3);
Mup, ys2, up95, lo95 = m3.predict( np.array(x) ); 

# Compute the values of the different AQs over the grid
mumax = max(Mup);
Z     = ( Mup - mumax - 0.01 ) / np.sqrt( ys2 )
PI    = sp.stats.norm.cdf( Z , 0, 1)
EI    = Z * sp.stats.norm.cdf( Z, 0, 1) + np.sqrt( ys2 ) * sp.stats.norm.pdf( Z, 0, 1)
UCB   = up95;

# Export data to R for plotting
out = np.hstack((Mup,up95,lo95,PI,EI,UCB));
pandas.DataFrame(out).to_csv("../figures/ch4-example-gp-acqfunc/ch4-example-gp-acqfunc-10d.csv");

########################################################################
# Export data to R for plotting
########################################################################

out = np.hstack((x,y));
pandas.DataFrame(out).to_csv("../figures/ch4-example-gp-acqfunc/ch4-example-gp-acqfunc-data.csv");

out = np.hstack((xred,yred));
pandas.DataFrame(out).to_csv("../figures/ch4-example-gp-acqfunc/ch4-example-gp-acqfunc-datared.csv");

########################################################################
# End of file
########################################################################